psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1600 @@
+    naux_ = auxiliary_->nbf();
+
+	p_shells_ = primary_->nshell();
+	Q_shells_ = auxiliary_->nshell();
+}
+
+//Used as testing machinery. It's nice to have a public function
+//to be able to look at private member data from pyside.
+//I plan on nuking this as soon as I get this project done,
+//don't you worry. JSOB
+
+//syntax copied and adapted from DiskDFJK.
+//I decided to forgo copying from MemDFJK because
+//that code seems more specialized. 
+void DirectDFJK::print_header() const {
+	if (print_)	{
+		outfile->Printf("==> DirectDFJK: Density-Fitted J/K Matrices <==\n\n");
+
+        outfile->Printf("    J tasked:          %11s\n", (do_J_ ? "Yes" : "No"));
+        outfile->Printf("    K tasked:          %11s\n", (do_K_ ? "Yes" : "No"));
+		outfile->Printf("    wK tasked:         %11s\n", (do_wK_ ? "Yes" : "No"));
+		if (do_wK_) outfile->Printf("Don't interpret these results.\nWe don't have w integrals in DirectDFJK\n");
+		outfile->Printf("    OpenMP threads:    %11d\n", omp_nthread_);
+		outfile->Printf("    Integrals threads: %11d\n", df_ints_num_threads_);
+		outfile->Printf("    Memory [MiB]:      %11ld\n", (memory_ * 8L) / (1024L*1024L));
+		outfile->Printf("    Algorithm:         %11s\n", "Direct");
+        outfile->Printf("    Schwarz Cutoff:    %11.0E\n", cutoff_);
+        outfile->Printf("    Fitting Condition: %11.0E\n\n", condition_);
+
+        outfile->Printf("   => Auxiliary Basis Set <=\n\n");
+        auxiliary_->print_by_level("outfile", print_);
+	}
+}
+
+// In the long term, I'm not sure what belongs here and what
+// belongs in common_init() JSOB
+
+// I think this should call the memory estimate and determine the number of blocks.
+// Memory will be allocated locally by the compute_JK() method.
+
+void DirectDFJK::preiterations() {
+
+    //cutoff_ = 1e-12;
+	// set_df_ints_num_threads(omp_nthread_);
+	sparsity_prep_pQq();
+
+	//prepare blocks tells us at which indices we start and stop each block.
+	// I'm making it a separate function because I've seen evidence that 
+	// DFHelper does some resizing, and I want to be able to match that.
+//dirty
+	biggest_shell_ = 0;
+	for (int i = 0; i < primary_->nshell(); i++){
+		if (primary_->shell(i).nfunction() > biggest_shell_) {
+			biggest_shell_ = primary_->shell(i).nfunction();
+		}
+	}
+//end dirty
+
+	// prepares the coulomb metric. Of course, our calculations depend on the
+	//  inverse of the coulomb metric and the (-0.5) power of the metric, so
+	//  we will prepare those immediately afterwards
+	//prepare_metric();
+	get_met();
+	
+	prepare_metric_power(-0.5);
+
+	//for what it's worth, we should have the density matrix
+	//prepare_D_symm();
+
+}
+
+void DirectDFJK::sparsity_prep_pQq(){
+	int procs = 1;
+#ifdef _OPENMP
+	procs = omp_nthread_;
+#endif
+
+	// variables to hold data
+	double global_max_int = 0.0;
+
+
+	std::unique_ptr<double[]> shel_maxes(new double[p_shells_*p_shells_]);
+	std::unique_ptr<double[]> integrals(new double[nbf_*nbf_]);
```

Comment:
Why not `std::vector<double>`?

Reply:
I didn't plan on resizing the data, so I thought unique_ptr would be sufficient.

Reply:
Then use `std::array`. `new` and `malloc` are things to avoid, since there's no performance benefit but debugging is harder;  you just make it easier to shoot yourself in the foot.

Reply:
As far as I can tell, `std::array` requires knowing the size of the data at compile time which wouldn't be true about this case because we rely on the size of the basis set here. Otherwise, I would gladly use it.

I'm also happy to use the standard libraries to manage allocation and deallocation of memory in my code. That is what motivated my use of smart pointers as you can see in lines 157 and 158. I'm not a huge fan of how I have to put `new` in there myself, but I did a lot of validation to make sure this was right.

Reply:
If `std::array` doesn't work, then use `std::vector`. There is no performance hit; it just allows for cleaner code i.e. no `new` or `free`. This is already preferred elsewhere in Psi

Reply:
`std::unique_ptr` also does not require use of free. There is also preference for `std::unique_ptr` elsewhere in psi

Reply:
I'm still confused about why you're bringing up `new` and `delete` because the `std::unique_ptr` takes care of those operations.

Beyond the point of `std::unique_ptr` addressing concerns about memory management and smart pointers in general being useful, lightweight, and a part of several key pieces of code in Psi, I would have to change all the memory accessing in the code which could be problematic for a number of reasons:

1) It would add no enhancement to functionality.
2) It would require changes to several hundred lines of code which I would need to do between now and December 31 when I am off Georgia Tech's payroll.

Reply:
Sorry, I just saw the `new` statements, I didn't realize you had `unique_ptr`.